src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -63,6 +63,13 @@ internal partial class AnalysisState
         private readonly ObjectPool<HashSet<CompilationEvent>> _compilationEventsPool;
         private readonly HashSet<CompilationEvent> _pooledEventsWithAnyActionsSet;
 
+        // Create static pools for heavily allocated per-analyzer state objects - this helps in reducing allocations across CompilationWithAnalyzer instances.
+        private const int PoolSize = 1000;
+        private static readonly ObjectPool<AnalyzerStateData> s_analyzerStateDataPool = new ObjectPool<AnalyzerStateData>(() => new AnalyzerStateData(), PoolSize);
```

Comment:
due to that reason, if you see any pool in IDE (since it showed up in various cases), for any collections, it checks size before putting it back to pool and if it became way big than expected, it either throw that object away or shrink it.